### ISSUE TITLE
Pebble is not always waiting for finished processes

### ISSUE THREAD

**Issue by phvalguima** [canonical/pebble#71]
There is [pebble#6](https://github.com/canonical/pebble/issues/6) where it raises how pebble should be working with system signals.

I've noticed that, over time, zombie processes start to show up in my app container.

I am running on top of a Ubuntu:20.04 image, and I can get a prompt on my container and see:
```
# ps aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.0  0.0 713020 10452 ?        Ssl  10:04   0:00 /charm/bin/pebble run --create-dirs --hold --verbose
postgres      58  0.0  0.0      0     0 ?        Zs   10:04   0:00 [postgres] <defunct>
postgres      93  0.0  0.1 215584 27032 ?        S    10:04   0:00 postgres
postgres     114  0.0  0.0 215696  8148 ?        Ss   10:04   0:00 postgres: checkpointer   
postgres     115  0.0  0.0 215584  5556 ?        Ss   10:04   0:00 postgres: background writer   
postgres     116  0.0  0.0 215584  9772 ?        Ss   10:04   0:00 postgres: walwriter   
postgres     117  0.0  0.0 216140  8428 ?        Ss   10:04   0:00 postgres: autovacuum launcher   
postgres     118  0.0  0.0  69948  4540 ?        Ss   10:04   0:00 postgres: archiver   
postgres     119  0.0  0.0  69948  4540 ?        Ss   10:04   0:00 postgres: stats collector   
postgres     120  0.0  0.0 216004  6516 ?        Ss   10:04   0:00 postgres: logical replication launcher   
root         137  0.0  0.0   7244  3812 pts/0    Ss   10:16   0:00 bash
root         144  0.0  0.0   8900  3232 pts/0    R+   10:16   0:00 ps aux
```

Checking for the parent PID and all of them have pebble (PID 1) as its parent.

So, I collected some straces and I noticed that pebble does treat SIGCHLD in some cases:
```
[pid 1294529] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=207, si_uid=999, si_status=0, si_utime=5, si_stime=8} ---
[pid 1295049] futex(0xc78eb8, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 1294529] rt_sigreturn({mask=[]} <unfinished ...>
[pid 1295049] <... futex resumed>)      = 1
[pid 1294529] <... rt_sigreturn resumed>) = 1
[pid 1294524] <... futex resumed>)      = 0
[pid 1295049] wait4(207,  <unfinished ...>
```

However, there are other occasions where the SIGCHLD seems to be ignored. If I run strace early enough on the pebble process, I can see that at the beginning, SIGCHLD will be ignored.
```
[pid 1404940] 16:06:59 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=273, si_uid=999, si_status=1, si_utime=2, si_stime=5} ---
[pid 1414197] 16:06:59 clone(child_stack=NULL, flags=CLONE_VM|CLONE_VFORK|SIGCHLDstrace: Process 1414287 attached
[pid 1414287] 16:06:59 rt_sigaction(SIGCHLD, {sa_handler=SIG_DFL, sa_mask=~[], sa_flags=SA_RESTORER|SA_ONSTACK|SA_RESTART|SA_SIGINFO, sa_restorer=0x465190}, NULL, 8) = 0
[pid 1414287] 16:06:59 rt_sigaction(SIGCHLD, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f66aeeff210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
[pid 1414287] 16:06:59 rt_sigaction(SIGCHLD, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f66aeeff210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f66aeeff210}, 8) = 0
[pid 1414287] 16:06:59 rt_sigaction(SIGCHLD, {sa_handler=0x55bbedb74aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f66aeeff210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f66aeeff210}, 8) = 0
[pid 1414287] 16:06:59 clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLDstrace: Process 1414288 attached
[pid 1414287] 16:06:59 rt_sigaction(SIGCHLD, {sa_handler=0x55bbedb74aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f66aeeff210},  <unfinished ...>
[pid 1414288] 16:06:59 rt_sigaction(SIGCHLD, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f66aeeff210}, {sa_handler=0x55bbedb74aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f66aeeff210}, 8) = 0
[pid 1414288] 16:06:59 rt_sigaction(SIGCHLD, {sa_handler=0x55bbedb74aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f66aeeff210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f66aeeff210}, 8) = 0
[pid 1414288] 16:06:59 rt_sigaction(SIGCHLD, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f66aeeff210}, {sa_handler=0x55bbedb74aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f66aeeff210}, 8) = 0
```

**Comment by benhoyt** [comment]
Thanks for the report, @phvalguima. I'm not quite sure what the request is in this issue, though. Will this not be fixed the issue you've linked, #6?

**Comment by benhoyt** [comment]
This should be fixed by #101 (just merged). Note that there's also https://github.com/canonical/pebble/issues/6, which will be fixed by https://github.com/canonical/pebble/pull/114.
